### PR TITLE
Architecture and design for implementation of a handler for consuming "composite" formats

### PR DESCRIPTION
**Pull request by danielgrant**

### CODE REVIEW

**Review by tomcrane — COMMENTED** [review]
In the flow chart, the incoming payload might be a Collection (POST to a queue) or an individual asset (PUT to location). The decision to invoke the PDF handler would be based on whether the member(s) of the Collection, or the single PUTted asset, are of type vocab:Composite.
This is slightly different to the flow in the diagram I think.
We can have an extra condition for this work - you can't mix composites with other types in a single API call. You can't post a collection to a queue that has an image and a composite in it. Do that in two separate calls please!

**Comment by danielgrant** [comment]
> In the flow chart, the incoming payload might be a Collection (POST to a queue) or an individual asset (PUT to location). The decision to invoke the PDF handler would be based on whether the member(s) of the Collection, or the single PUTted asset, are of type vocab:Composite. This is slightly different to the flow in the diagram I think. We can have an extra condition for this work - you can't mix composites with other types in a single API call. You can't post a collection to a queue that has an image and a composite in it. Do that in two separate calls please!

I've updated this diagram to make it a bit simpler now by requiring the external system which invokes the API to call the Composite Handler directly, instead of having to go via the API. I believe this is what we discussed, and it removes the need to implement any modifications in that API component.

File: docs/adr/0001-composite-handler-design.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+- Artifact processing / rasterization code already highly optimised.
+
+#### Negative Consequences
+
+- Requires modification of multiple existing components, i.e.
+  Orchestrator, thumbnail generation, etc.
+- Potentially a heavyweight solution, Cantaloupe image server can be
+  resource intensive.
+- Results in two image servers deployed within the DLCS - IIPImage and
+  Cantaloupe - with signficant effort to remove IIPImage entirely.
```
Comment (stephenwf) [inline docs/adr/0001-composite-handler-design.md:69]: I guess vendor lock-in would also be a negative too

Comment (danielgrant) [reply docs/adr/0001-composite-handler-design.md]: Good point - updated.

File: docs/adr/0001-composite-handler-design.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+which can process "composite" artifacts. Such artifacts typically take
+the form of a single object containing multiple embedded resources, each
+of which can be rasterized and treated as an individual resource within
+the DLCS.
+
+## Considered Options
+
+- Additional standalone component which consumes a DLCS ingest payload,
+  retrieves the object, rasterizes accordingly, and then invokes the
+  existing DLCS ingestion workflow.
```
Comment (stephenwf) [inline docs/adr/0001-composite-handler-design.md:25]: Could the standalone tool use cantaloupe to rasterize? That could make it easier if we wanted to switch or had to switch down the line. Or does that defeat the purpose of having it standalone?

Comment (danielgrant) [reply docs/adr/0001-composite-handler-design.md:25]: When we discussed this, that was tabled as an option, but I think in doing so we end up doing more work than we have to - deployment, configuration and maintenance of Cantaloupe, in addition to the development of this additional standalone component which simply proxies through to that image server.

Contrast this with implementing rasterization within the handler - which _should_ be a relatively simple task - and to me I don't think it makes sense to do both.

Comment (tomcrane) [reply docs/adr/0001-composite-handler-design.md:25]: Yeah - we _could_ do this, but I think we'll come back to it later when we have Cantaloupe as a part of the DLCS.

There is a related issue here: https://github.com/dlcs/protagonist/issues/30#issuecomment-721274027

...where Cantaloupe might be used at ingest time as well as at end-user-facing runtime, to generate thumbs from sources without having to move them from S3.

**Comment by donaldgray** [comment]
LGTM. Agree this reflects what we discussed on the call. We could revisit and go via the API once we've ported it over to .net 5 (likely 6 by the time it's done) but what's in the flowchart will definitely work.

One observation on the flowchart - doesn't necessarily affect the viability of the suggested solution but could have an affect on longer term maintenance - is the _Extract the origin URI and retrieve the artifact..._ step. Currently the Engine component does this by using a set of know "Origin Strategies" (http, s3, ftp) and a DB lookup to determine which strategy to use. Adding logic to handle each of these to the CompositeHandler isn't particularly complex but does involve having that logic in multiple places and we'd need to keep them in sync. Until now the ingest process has been for the engine to move an artifact from Origin->LocalStorage then handing off to another component for processing. 

Given that this component is driven by a client requirement we could just implement that single strategy (likely S3) that they need and can expand later.

**Comment by danielgrant** [comment]
> One observation on the flowchart - doesn't necessarily affect the viability of the suggested solution but could have an affect on longer term maintenance - is the _Extract the origin URI and retrieve the artifact..._ step. Currently the Engine component does this by using a set of know "Origin Strategies" (http, s3, ftp) and a DB lookup to determine which strategy to use. Adding logic to handle each of these to the CompositeHandler isn't particularly complex but does involve having that logic in multiple places and we'd need to keep them in sync. Until now the ingest process has been for the engine to move an artifact from Origin->LocalStorage then handing off to another component for processing.
> 
> Given that this component is driven by a client requirement we could just implement that single strategy (likely S3) that they need and can expand later.

Good point. As you note, I think we can do one of two things here:

- Duplicate the origin strategies logic in this component - which requires DB interaction - to accommodate all potential strategies; or
- Implement the only strategy that we know we need to meet the specific client requirement that is driving this piece of work.

I'm including to do the latter, and kick the former into the long grass until the API is updated to accept composite requests. When we do this work, we could have the API handle retrieval of the artifact from the origin and place it into an accessible location within the DLCS, and then delegate the request to the CompositeHandler. This way, the API continues to do the "heavy lifting" of artifact retrieval, and there's no duplication of that logic.

Something to think about.

**Comment by tomcrane** [comment]
I'm OK to approve the RFC part of this PR once it's separate from the Composite handler implementation part.

On the format strings - it's easier to insist that the format string syntax is that of the code it's going to hit in the DLCS, which is now Python - so it's easier for us, the onus is on the caller to get it right rather than for us to transform it into a different language format.

And, it's probably more likely that clients will be scripting ingests in Python than in C#, so it's good from that POV.

Later, if/when we make the DLCS API accept and process the composite directly, we will need to transform the syntax to C# format strings. Maybe there's a nice library to do that (couldn't see one from a quick Google).

**Pull request closed without merging**